### PR TITLE
deprecate TLS_DH_* (and TLS_DHE_* in TLS 1.2) ...

### LINKED ISSUE #244: Ref to draft-aviram-tls-deprecate-obsolete-kex
Which is likely to be adopted by TLS WG.

### LINKED ISSUE #274: Harmonize with Recommendations in draft-tls-deprecate-{obsolete-kex,ffdhe}
Nimrod Aviram, co-author of draft-bartle-tls-deprecate-ffdhe, notes that here are a few significant differences between the recommendations in that draft and in draft-ietf-uta-7525bis:

- The draft-ietf-uta lists RSA key exchange as a SHOULD NOT. We've had similar discussions in the TLS WG, and I argue that RSA should be a MUST NOT. We've had support for this on the TLS WG.

- The wording in Section 4.1 of draft-ietf-uta implies that using finite field DHE cipher suites is generally good practice. Most web client implementations have dropped support for finite field DHE. Further, the Introduction of WIP draft-tls-deprecate-obsolete-kex lists problems affecting finite field DHE, especially when exponents are reused. These problems are arguably severe enough to make exponent reuse a MUST NOT. Section 6.4 has both static finite field DH and exponent reuse as a SHOULD NOT.

- On a side note, the list of recommended cipher suites in Section 4.2 is a subset of the recommended cipher suites in the "Intermediate" configuration in Mozilla's Server Side TLS Guide. Could one of the authors please explain the rationale for this difference?

Obviously, my recommendations are reflected in the WIP draft-tls-deprecate-obsolete-kex:
(please excuse the brevity)
- MUST NOT use (non-ephemeral) DH cipher suites.
- SHOULD NOT use non-ephemeral ECDH.
- Finite field DHE: MUST NOT reuse exponents, MUST use a well-known group.
- MUST NOT use RSA key exchange.

### LINKED ISSUE #289: Cite RFC 8422?
Should we cite RFC 8422 "Elliptic Curve Cryptography (ECC) Cipher Suites for Transport Layer Security (TLS) Versions 1.2 and Earlier"?

### PR DESCRIPTION
**Pull request by thomas-fossati**
... and also:
* s/RFC4492/RFC8422/ + allow ec_point_formats=uncompressed for backwards compat
* recommend X25519 alongside NIST P-256
* drop TLS_DHE_RSA_WITH_AES from the recommended ciphers in TLS 1.2
* make the case for ECDH over FFDH somewhat stronger

Note: this PR *does not* include the fix around DH public key reuse.

Fixes #244
Fixes #274
Fixes #289

### CODE REVIEW

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 1]
```diff
@@ -499,11 +499,21 @@ Cryptographic algorithms weaken over time as cryptanalysis improves: algorithms
                with the string "TLS_RSA_WITH_*", have several drawbacks, especially
                the fact that they do not support forward secrecy.
             
+* Implementations SHOULD NOT negotiate cipher suites based on
+               non-ephemeral finite-field Diffie-Hellman key agreement.
```
Comment (yaronf) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:503]:
```suggestion
               non-ephemeral (static) finite-field Diffie-Hellman key agreement.
```

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 2]
```diff
@@ -499,11 +499,21 @@ Cryptographic algorithms weaken over time as cryptanalysis improves: algorithms
...
+
+  Rationale: These cipher suites, which have assigned values starting
+               with the string "TLS_DH_*", have several drawbacks, especially
```
Comment (yaronf) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:506]:
```suggestion
  Rationale: These cipher suites, which have assigned values prefixed by "TLS_DH_*", have several drawbacks, especially
```

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 3]
```diff
@@ -499,11 +499,21 @@ Cryptographic algorithms weaken over time as cryptanalysis improves: algorithms
...
 
 * Implementations MUST support and prefer to negotiate cipher suites 
                offering forward secrecy, such as those in the Ephemeral 
                Diffie-Hellman and Elliptic Curve Ephemeral Diffie-Hellman ("DHE" 
-               and "ECDHE") families.
+               and "ECDHE") families.  However, given the known limitations
+               around both construction (e.g., {{RACCOON}}) and negotiation
+               (including using {{?RFC7919}}), TLS 1.2 implementations SHOULD
+               NOT negotiate cipher suites based on ephemeral finite-field
+               Diffie-Hellman key agreement (i.e., "TLS_DHE_*" suites).
```
Comment (yaronf) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:516]: Should we clarify what we don't like about 7919: is it still vulnerable? Is it just not well deployed?

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 4]
```diff
@@ -541,11 +547,11 @@ Servers MUST prefer this cipher suite over weaker cipher suites whenever it is p
...
+This document does not change the mandatory-to-implement (MTI) TLS cipher suite(s) prescribed by TLS. To maximize interoperability, RFC 5246 mandates implementation of the TLS_RSA_WITH_AES_128_CBC_SHA cipher suite, which is significantly weaker than the cipher suites recommended here. (The GCM mode does not suffer from the same weakness, caused by the order of MAC-then-Encrypt in TLS {{Krawczyk2001}}, since it uses an AEAD mode of operation.) Implementers should consider the interoperability gain against the loss in security when deploying the TLS_RSA_WITH_AES_128_CBC_SHA cipher suite. Other application protocols specify other cipher suites as MTI.
```
Comment (yaronf) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:550]: Maybe we should bite the bullet and declare `TLS_RSA_WITH_AES_128_CBC_SHA` dead. I think we don't even need to "update" 5246 because it's already obsolete...

This is out of scope for this PR though.

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 5]
```diff
@@ -541,11 +547,11 @@ Servers MUST prefer this cipher suite over weaker cipher suites whenever it is p
...
 
 Note that some profiles of TLS 1.2 use different cipher suites. For example, {{?RFC6460}} defines a profile that uses the TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 and TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 cipher suites.
 
-{{!RFC4492}} allows clients and servers to negotiate ECDH parameters (curves).  Both clients and servers SHOULD include the "Supported Elliptic Curves" extension {{!RFC4492}}.  For interoperability, clients and servers SHOULD support the NIST P-256 (secp256r1) curve {{!RFC4492}}. In addition, clients SHOULD send an ec_point_formats extension with a single element, "uncompressed".
+{{!RFC8422}} allows clients and servers to negotiate ECDH parameters (curves).  Both clients and servers SHOULD include the "Supported Elliptic Curves" extension {{!RFC8422}}.  Clients and servers SHOULD support the NIST P-256 (secp256r1) curve {{!RFC8422}} and X25519 {{!RFC7748}}. For backwards compatibility purposes, clients MAY send an ec_point_formats extension with a single element, "uncompressed".
```
Comment (yaronf) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:554]: Why did you change the SHOULD on ec_point_formats to MAY?

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 6]
```diff
@@ -765,7 +771,7 @@ In the context of TLS and DTLS, such compromise of long-term keys is not entirel
...
+Forward secrecy is generally achieved by using the Diffie-Hellman scheme to derive session keys. The Diffie-Hellman scheme has both parties maintain private secrets and send parameters over the network as modular powers over certain cyclic groups. The properties of the so-called Discrete Logarithm Problem (DLP) allow the parties to derive the session keys without an eavesdropper being able to do so. There is currently no known attack against DLP if sufficiently large parameters are chosen. A variant of the Diffie-Hellman scheme uses elliptic curves instead of the originally proposed modular arithmetic. Given the current state of the art, elliptic-curve Diffie-Hellman appear to be more efficient, permit shorter key lengths, and allow less freedom for implementation errors than finite-field Diffie–Hellman.
```
Comment (yaronf) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:774]:
```suggestion
Forward secrecy is generally achieved by using the Diffie-Hellman scheme to derive session keys. The Diffie-Hellman scheme has both parties maintain private secrets and send parameters over the network as modular powers over certain cyclic groups. The properties of the so-called Discrete Logarithm Problem (DLP) allow the parties to derive the session keys without an eavesdropper being able to do so. There is currently no known attack against DLP if sufficiently large parameters are chosen. A variant of the Diffie-Hellman scheme uses elliptic curves instead of the originally proposed modular arithmetic. Given the current state of the art, elliptic-curve Diffie-Hellman appears to be more efficient, permits shorter key lengths, and allows less freedom for implementation errors than finite-field Diffie–Hellman.
```

**Review by yaronf — CHANGES_REQUESTED** [review]

File: BCP195bis/draft-ietf-uta-rfc7525bis.md (hunk 3 above)
Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md:515]: explicitly state the reasons why it is not widely implemented / not usable in 1.2

File: BCP195bis/draft-ietf-uta-rfc7525bis.md (hunk 4 above)
Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md:549]: This is issue #292 now

File: BCP195bis/draft-ietf-uta-rfc7525bis.md (hunk 5 above)
Comment (stpeter) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md:553]: We could say "but this is not strictly necessary because "uncompressed" is the only defined value for this extension".

**Comment by thomas-fossati** [comment]
@thomas-fossati ACTION: Add the changes to the log

File: BCP195bis/draft-ietf-uta-rfc7525bis.md (hunk 5 above)
Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md:553]: great, will add.

Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md]: I have reworded the paragraph in 96994b9

**Comment by thomas-fossati** [comment]
> @thomas-fossati ACTION: Add the changes to the log

addressed in cd8efeb

File: BCP195bis/draft-ietf-uta-rfc7525bis.md (hunk 3 above)
Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md]: See 5f5c461

**Review by stpeter — APPROVED** [review]

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 7]
```diff
@@ -541,11 +547,11 @@ Servers MUST prefer this cipher suite over weaker cipher suites whenever it is p
...
 
 Note that some profiles of TLS 1.2 use different cipher suites. For example, {{?RFC6460}} defines a profile that uses the TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 and TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 cipher suites.
 
-{{!RFC4492}} allows clients and servers to negotiate ECDH parameters (curves).  Both clients and servers SHOULD include the "Supported Elliptic Curves" extension {{!RFC4492}}.  For interoperability, clients and servers SHOULD support the NIST P-256 (secp256r1) curve {{!RFC4492}}. In addition, clients SHOULD send an ec_point_formats extension with a single element, "uncompressed".
+{{!RFC8422}} allows clients and servers to negotiate ECDH parameters (curves).  Both clients and servers SHOULD include the "Supported Elliptic Curves" extension {{!RFC8422}}.  Clients and servers SHOULD support the NIST P-256 (secp256r1) {{!RFC8422}} and X25519 (x25519) {{!RFC7748}} curves.  Note that {{!RFC8422}} deprecates all but the uncompressed point format.  Therefore, if the client sends an ec_point_formats extension, the ECPointFormatList MUST contain a single element, "uncompressed".
```
Comment (yaronf) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:554]: See new discussion on CFRG, https://mailarchive.ietf.org/arch/msg/cfrg/Ev8hgyojKeObXMZ7SF2m3_yekMo/

I believe it only applies to EdDSA certificates/signatures, not to ECDH, correct?

We should also clarify this point here, I think.

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 8]
```diff
@@ -499,11 +499,21 @@ Cryptographic algorithms weaken over time as cryptanalysis improves: algorithms
...
 * Implementations MUST support and prefer to negotiate cipher suites 
                offering forward secrecy, such as those in the Ephemeral 
                Diffie-Hellman and Elliptic Curve Ephemeral Diffie-Hellman ("DHE" 
-               and "ECDHE") families.
+               and "ECDHE") families.  However, given the known fragility
+               of the construction (see {{RACCOON}}) and the limitation around
+               negotiation (including using {{?RFC7919}}, which has seen very
+               limited uptake), TLS 1.2 implementations SHOULD NOT negotiate
+               cipher suites based on ephemeral finite-field Diffie-Hellman key
+               agreement (i.e., "TLS_DHE_*" suites).
```
Comment (yaronf) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:516]: This bullet has become confusing. I suggest to reword and remove DHE from the MUST sentence at the top.

File: BCP195bis/draft-ietf-uta-rfc7525bis.md (hunk 7 above)
Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md:554]: > See new discussion on CFRG, https://mailarchive.ietf.org/arch/msg/cfrg/Ev8hgyojKeObXMZ7SF2m3_yekMo/

Thanks for the pointer.

> I believe it only applies to EdDSA certificates/signatures, not to ECDH, correct?

Yes, it's correct.

> We should also clarify this point here, I think.

We chose to not recommend ECDSA ciphersuites -- although [it looks like they're not so unpopular](https://notary.icsi.berkeley.edu/#certificate-public-key-details) -- so I don't know exactly where to anchor a clarification.

File: BCP195bis/draft-ietf-uta-rfc7525bis.md (hunk 8 above)
Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md:514]: Done in 66c60e0

File: BCP195bis/draft-ietf-uta-rfc7525bis.md (hunk 7 above)
Comment (yaronf) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md:552]: Hmmm, the ICSI reference is quite old actually, but here's [newer information](https://www.f5.com/labs/articles/threat-intelligence/the-2021-tls-telemetry-report), "The move to elliptic curve cryptography is slow but steady, with 25 percent of certificates now signed with the Elliptic Curve Digital Signature Algorithm (ECDSA)..." - I think we need to reopen that discussion.

Comment (thomas-fossati) [reply BCP195bis/draft-ietf-uta-rfc7525bis.md:552]: See #218

File: BCP195bis/draft-ietf-uta-rfc7525bis.md [hunk 9]
```diff
@@ -499,11 +499,19 @@ Cryptographic algorithms weaken over time as cryptanalysis improves: algorithms
...
 * Implementations MUST support and prefer to negotiate cipher suites 
-               offering forward secrecy, such as those in the Ephemeral 
-               Diffie-Hellman and Elliptic Curve Ephemeral Diffie-Hellman ("DHE" 
-               and "ECDHE") families.
+               offering forward secrecy.  However, given the known fragility
+               of the construction (see {{RACCOON}}) and the limitation around
+               negotiation (including using {{?RFC7919}}, which has seen very
+               limited uptake), TLS 1.2 implementations SHOULD NOT negotiate
+               cipher suites based on ephemeral finite-field Diffie-Hellman key
+               agreement (i.e., "TLS_DHE_*" suites).
```
Comment (thomas-fossati) [inline BCP195bis/draft-ietf-uta-rfc7525bis.md:514]:
```suggestion
               offering forward secrecy.  However, TLS 1.2 implementations SHOULD NOT negotiate
               cipher suites based on ephemeral finite-field Diffie-Hellman key
               agreement (i.e., "TLS_DHE_*" suites), given the known fragility
               of the construction (see {{RACCOON}}) and the limitation around
               negotiation. including using {{?RFC7919}}, which has seen very
               limited uptake.
```
